src/golem-network.ts

```diff
@@ -251,9 +259,13 @@ export class GolemNetwork {
     const agreement = await this.market.proposeAgreement(draftProposal);
     proposalSubscription.unsubscribe();
 
-    const allocation = await this.payment.createAllocation({ budget: 1 });
     const lease = this.market.createLease(agreement, allocation);
 
+    // Attach handlers for cleanup after all work's done
+    lease.events.once("finalized", async () => {
+      await this.payment.releaseAllocation(allocation);
+    });
```

Comment:
If the lease fails to finalize (for example provider never sends the invoice or there is some error while killing activity) then theoretically the allocation will never be cleared :/ Not sure what to do about this tbh, maybe `finalized` should emit even if the function errors

Reply:
I will cover this and trigger the event in `finally`